R/restopt_problem.R

```diff
@@ -0,0 +1,87 @@
+#' @include internal>R
+NULL
+
+# Copyright (c) 2021, Dimitri Justeau-Allaire
+#
+# Institut Agronomique neo-Caledonien (IAC), 98800 Noumea, New Caledonia
+# AMAP, Univ Montpellier, CIRAD, CNRS, INRA, IRD, Montpellier, France
+#
+# This file is part of restoptr
+#
+# restoptr is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# restoptr is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with restoptr  If not, see <https://www.gnu.org/licenses/>.
+
+#' Restoration optimization problem
+#'
+#' Create a new restoration optimization problem using data that describe
+#' the spatial distribution of existing habitat and potential for
+#' restoration.
+#'
+#' @param existing_habitat [terra::rast()] Raster object containing binary
+#' values that indicate if each planning unit contains habitat or not.
+#'
+#' @param restorable_habitat [terra::rast()] Raster object containing
+#' continuous values that indicate the amount of habitat that can
+#' be restored within each planning unit.
+#'
+#' @return A new restoration problem (`RestoptProblem`) object.
+#'
+#' @details
+#' TODO.
+#'
+#' @examples
+#' # TODO
+#'
+#' @export
+restopt_problem <- function(existing_habitat, restorable_habitat) {
+  # assert arguments are valid
+  ## initial checks
+  assertthat::assert_that(
+    inherits(existing_habitat, "SpatRaster"),
+    inherits(restorable_habitat, "SpatRaster")
+  )
+  ## further checks
+  assertthat::assert_that(
+    terra::has_values(existing_habitat),
+    terra::has_values(restorable_habitat),
+    terra::compareGeom(
+      existing_habitat, restorable_habitat, stopiffalse = FALSE
+    )
+  )
+
+  # return object
```

Comment:
If I understand correctly, the initial S4 class would become a simpler data frame, that would be processed in the `restopt_compile` function? I'm ok to go with that, as it seems simple and efficient!

Reply:
Yeah, that's what I was thinking. The `restopt_problem()` function would create an S3 class object (basically a list with a defined class) that could be updated using objective/constraint functions. After the user has finished specifying their problem, they would call `solve()` to (1) build the problem specification through the Java API (via the `compile()` function), (2) run the optimization, and (3) return optimization results (e.g. as a `terra::rast()` object).

Reply:
Ah - I just noticed you described these exact steps in a later comment

Reply:
I think we're in agrement, so I'll mark this comment thread as resolved - but please feel free to re-open if you want to chat more about this.